src/Data/Text/Utf8.hs

```diff
@@ -396,3 +340,94 @@ unsafeCutUtf8 (CodeUnitIndex !begin) (CodeUnitIndex !length) (Text !u8data !off
 unsafeSliceUtf8 :: CodeUnitIndex -> CodeUnitIndex -> Text -> Text
 unsafeSliceUtf8 (CodeUnitIndex !begin) (CodeUnitIndex !length) (Text !u8data !off !_len) =
   Text u8data (off + begin) length
+
+-- $generalFunctions
+--
+-- These functions are available in @text@ as well and should be removed once this library moves to @text-2@.
+-- You should be able to use these by doing @import qualified Data.Text.Utf8 as Text@ just like you would with @text@.
+--
+-- NOTE: The 'Text' instances for @Show@, @Eq@, @Ord@, @IsString@, @FromJSON@, @ToJSON@ and @Hashable@ in this file also fall in this category.
+
+-- | TODO: Inefficient placeholder implementation.
+concat :: [Text] -> Text
+concat = pack . concatMap unpack
+
+-- | See 'Data.Text.dropWhile'.
+dropWhile :: (Char -> Bool) -> Text -> Text
+dropWhile predicate text =
+  let
+    len = codeUnitIndex (lengthUtf8 text)
+    go i
+      | i >= CodeUnitIndex len = i
+      | otherwise =
+        let
+          (codeUnits, codePoint) = unsafeIndexCodePoint text i
+        in
+          if predicate codePoint then
+            go $ i + codeUnits
+          else
+            i
+
+    prefixEnd = go 0
+  in
+    unsafeSliceUtf8 prefixEnd (CodeUnitIndex len - prefixEnd) text
+
+-- | Checks whether a text is the empty string.
+null :: Text -> Bool
+null (Text _ _ len) = len == 0
+
+-- | TODO: Inefficient placeholder implementation.
+pack :: String -> Text
+pack = go . stringToByteArray
+  where
+    go !arr = Text arr 0 $ sizeofByteArray arr
+
+-- | TODO: Inefficient placeholder implementation.
+-- See 'Data.Text.replicate'
+replicate :: Int -> Text -> Text
+replicate n = pack . Prelude.concat . Prelude.replicate n . unpack
+
+-- | TODO: Inefficient placeholder implementation.
+-- This function implements very basic string search. It's @text@ counterpart is 'Data.Text.Internal.Search.indices', which implements the Boyer-Moore algorithm.
+-- Since we have this function only to check whether our own Boyer-Moore implementation works, it would not make much sense to implement it using the same algorithm.
+-- Once we can use @text-2@, we can compare our implementation to the official @text@ one which presumably works.
+indices :: Text -> Text -> [Int]
+indices needle haystack
+  | needleLen == 0 = []
+  | otherwise = go 0 0
+  where
+    needleLen = lengthUtf8 needle
+    haystackLen = lengthUtf8 haystack
+
+    go startIdx needleIdx
+      -- needle is longer than remaining haystack
+      | startIdx + needleLen > haystackLen = []
+      -- whole needle matched
+      | needleIdx >= needleLen = codeUnitIndex startIdx : go (startIdx + needleLen) 0
+      -- charachter mismatch
+      | needleCp /= haystackCp = go (startIdx + 1) 0
+      -- advance
+      | otherwise = go startIdx $ needleIdx + codeUnits
+      where
+        (codeUnits, needleCp) = unsafeIndexCodePoint needle needleIdx
+        (_, haystackCp) = unsafeIndexCodePoint haystack $ startIdx + needleIdx
```

Comment:
Since we only use it for testing, would it make sense to move it to the test module?

Reply:
Sure, the idea behind putting it here is that this whole section (starting at line 344) can be removed when we move to `text-2` and no other changes have to be made in the testsuite.

Reply:
Ah, makes sense.